proposals/nnnn-strict-memory-safety.md

```diff
@@ -0,0 +1,536 @@
+
+Providing memory safety does not imply the absence of run-time failures. Good language design often means defining away runtime failures in the type system. However, memory safety requires only that an error in the program cannot be escalated into a violation of one of the safety properties. For example, having reference types be non-nullable by default defines away most problems with NULL pointers. With explicit optional types, the force-unwrap operator (postfix `!` ) meets the definition of memory safety by trapping at runtime if the unwrapped optional is `nil` . The standard library also provides the [`unsafelyUnwrapped` property](https://developer.apple.com/documentation/swift/optional/unsafelyunwrapped) that does not check for `nil` in release builds: this does not meet the definition of memory safety because it admits violations of initialization and lifetime safety that could be exploited.
+
+## Proposed solution
+
+This proposal introduces an opt-in strict memory safety checking mode that identifies all uses of unsafe behavior within the given module. There are several parts to this change:
+
+* A compiler flag `-strict-memory-safety` that enables warnings for all uses of unsafe constructs within a given module. All warnings will be in the diagnostic group `Unsafe`, enabling precise control over memory-safety-related warnings per [SE-0443](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0443-warning-control-flags.md).
+* An attribute `@unsafe` that indicates that a declaration is unsafe to use. Such declarations may use unsafe constructs within their signatures.
+* An `unsafe` expression that marks any use of unsafe constructs in an expression, much like `try` and `await`.
+* Standard library annotations to identify unsafe declarations.
+
+### Example of `unsafe` usage
+
+The `UnsafeBufferPointer` type will be marked with `@unsafe` in the Standard library, as will the other unsafe types (e.g., `UnsafePointer`, `UnsafeRawPointer`):
+
+```swift
+@unsafe 
+public struct UnsafeBufferPointer<Element> { ... }
+```
+
+This indicates that use of this type is not memory-safe. Any declaration that has `UnsafeBufferPointer` as part of its type is also unsafe, and would produce a warning under this strict safety mode, e.g.,
+
+```swift
+extension Array {
+  // warning on next line: reference to unsafe generic struct 'UnsafeBufferPointer'
+  func withUnsafeBufferPointerSimplified<T>(_ body: (UnsafeBufferPointer<Element>) -> T) -> T {
+    // ...
+  }
+}
+```
+
+This warning can be suppressed by marking the function as `@unsafe`:
+
+```swift
+extension Array {
+  @unsafe
+  func withUnsafeBufferPointerSimplified<T>(_ body: (UnsafeBufferPointer<Element>) -> T) -> T {
+    // ...
+  }
+}
+```
+
+Users of this function that also enable strict safety checking will see warnings when using it. For example:
```

Comment:
Would it be worth spelling out that adding `@unsafe` to a type or function has no impact on users that aren't using the mode?

Reply:
Sure